[content-visibility] Support content-visibility for top layers/dialogs

rwlbuis opened the pull request:
#### 5bb45c7cb923b098864c399395e128beef423901
<pre>
[content-visibility] Support content-visibility for top layers/dialogs
<a href="https://bugs.webkit.org/show_bug.cgi?id=245775">https://bugs.webkit.org/show_bug.cgi?id=245775</a>

Reviewed by Antti Koivisto.

Support ccontent-visibility for top layers/dialogs by not generating boxes [1]
for the dialog and its descendents if they are skipped content
because of content-visibility.

[1] <a href="https://w3c.github.io/csswg-drafts/css-contain/#cv-notes">https://w3c.github.io/csswg-drafts/css-contain/#cv-notes</a> [11]

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: <a href="https://commits.webkit.org/257646@main">https://commits.webkit.org/257646@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c554cea6c19d7f9995c729ad4185b36665f165c6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/99501 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32593 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108866 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169096 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/103493 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/9257 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85984 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91972 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106782 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105258 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6997 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90523 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33961 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88815 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21872 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/77239 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2526 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23387 "Passed tests") | | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2453 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45782 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/8606 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42865 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5263 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4313 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderElement.h:
```diff
@@ -409,6 +409,8 @@ class RenderElement : public RenderObject {
     void updateReferencedSVGResources();
     void clearReferencedSVGResources();
 
+    bool isTopLayerAncestor();
```
should be `const`. The name makes me pause a bit; perhaps clearer as `isAncestorOfTopLayerElement`

rwlbuis replied:
Right I missed that, new patch removes this code now.

smfr commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -854,6 +866,9 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         }
 
+        if (m_style.contentVisibility() != newStyle.contentVisibility() && isTopLayerAncestor())
+            document().renderView()->layer()->dirtyZOrderLists();
```
Seems pretty random. Would be clearer if the invalidation had some name that related to top layer things.

rwlbuis replied:
Indeed this was not nice, new patch removes this code luckily.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -3873,7 +3873,7 @@ Vector<RenderLayer*> RenderLayer::topLayerRenderLayers(const RenderView& renderV
     for (auto& element : renderView.document().topLayerElements()) {
         auto* renderer = element->renderer();
-        if (!renderer)
+        if (!renderer || renderer->isSkippedContent())
```
This is not the right way to fix this. The correct approach is to have RenderTreeUpdater just not generate boxes in this case.

nt1m replied:
We think it is the case because the spec mentions:

> If an element has an ancestor with [content-visibility: hidden](https://w3c.github.io/csswg-drafts/css-contain/#propdef-content-visibility), and it is placed in the [top layer](https://fullscreen.spec.whatwg.org/#top-layer), it does not generate any boxes, as if it were [display: none](https://w3c.github.io/csswg-drafts/css-display-3/#propdef-display).

rwlbuis replied:
A very good point, I missed that completely! It makes the patch much smaller too.

anttijk commented on Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
```diff
@@ -314,6 +314,9 @@ void RenderTreeUpdater::updateElementRenderer(Element& element, const Style::Ele
     bool shouldTearDownRenderers = elementUpdate.change == Style::Change::Renderer && (element.renderer() || element.hasDisplayContents() || element.displayContentsChanged());
+    bool isTopLayerHiddenByContentVisibility = element.isInTopLayer() && elementUpdate.style->effectiveSkipsContent() && renderTreePosition().parent().style().effectiveSkipsContent();
+    if (!shouldTearDownRenderers && isTopLayerHiddenByContentVisibility)
+        shouldTearDownRenderers = true;
```
This is probably not the right place to do this. Would testing in `Element::rendererIsNeeded` give the right behavior?

rwlbuis replied:
I tried, but unfortunately I think rendererIsNeeded would get too complicated or even couldn't do it (need to take parent/parent style into account). The problem is that we have two situations:
1. The dialog has content-visibility: hidden set on it, this means it should render but not its children.
2. The dialog is in a content-visibility: hidden subtree and in this case there is no point in even creating a renderer for it.
I did notice that my previous patch did too much work for the dialog children when content-visibility: hidden applies, so I adjusted the mayHaveRenderedDescendants logic. I think overall RenderTreeUpdater has the best info to make the right decision about content-visibility: hidden and dialogs.

anttijk reviewed (approved):
ok, lets see how this goes

rwlbuis committed:
[content-visibility] Support content-visibility for top layers/dialogs
https://bugs.webkit.org/show_bug.cgi?id=245775

Reviewed by Antti Koivisto.

Support ccontent-visibility for top layers/dialogs by not generating boxes [1]
for the dialog and its descendents if they are skipped content
because of content-visibility.

[1] https://w3c.github.io/csswg-drafts/css-contain/#cv-notes [11]

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::updateElementRenderer):

Canonical link: https://commits.webkit.org/257646@main
